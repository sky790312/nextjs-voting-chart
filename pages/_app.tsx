import { GlobalStyles, theme } from '@/GlobalStyles'
import { ThemeProvider } from 'styled-components'
import dayjs from 'dayjs'
import localizedFormat from 'dayjs/plugin/localizedFormat'
dayjs.extend(localizedFormat)

export default function App({ Component }) {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Component />
    </ThemeProvider>
  )
}
