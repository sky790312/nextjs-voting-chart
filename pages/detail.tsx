import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import Head from 'next/head'
import { fetchPoll } from '@/api'
import { PoolSchema } from '@/api/schema'
import dayjs from 'dayjs'
import { Doughnut } from 'react-chartjs-2'
import { useRouter } from 'next/router'
import { getDonutChartData } from '@/utils'
import { Container } from '@/GlobalStyles'

export const Detail = React.memo(
  (): JSX.Element => {
    const router = useRouter()
    const pollId = router.query.id
    const [poll, setPoll] = useState<PoolSchema>()
    // const [userAnswer, setUserAnswer] = useState<string>()

    const donutChartConfig = {
      labels: poll?.answer.options.map(option => option.label),
      data: poll?.answer.options.map((option, index) => (index + 1) * 10),
    }
    const data = getDonutChartData(donutChartConfig)

    // TODO: fake data, should using poll answer data.
    const total = data?.datasets[0].data?.reduce((acc, cur) => {
      acc += cur
      return acc
    }, 0)

    useEffect(() => {
      const fetchData = async () => {
        try {
          const data = await fetchPoll(Number(pollId))
          setPoll(data)
        } catch (error) {
          console.log(error)
        }
      }

      fetchData()
    }, [pollId])

    // useEffect(() => {
    //   // const foundAnswerIndex =  poll?.answer.options.findIndex(answer => answer.label === userAnswer)
    //   // TODO: here should update poll data's answer value
    // // eslint-disable-next-line react-hooks/exhaustive-deps
    // }, [userAnswer])

    const handleAnswer = e => {
      // const answer = e.target.value
      // setUserAnswer(answer)
      // TODO: here should update poll data's answer value, not using fake userAnwser and it will rerender chart and total votes.
    }

    return (
      <>
        <Head>
          <title>Create Next App</title>
          <link rel='icon' href='/favicon.ico' />
        </Head>
        <DetailContainer>
          <div>
            <h1>{poll?.title}</h1>
            <StyledPublishedDate>
              Publish: {dayjs(poll?.publishedDate).format('LLLL')}
            </StyledPublishedDate>
            <div>
              {poll?.answer.options.map(option => (
                <span key={option.id}>
                  <input
                    type={poll?.answer.type === 'Single' ? 'radio' : 'checkbox'}
                    id={String(option.id)}
                    value={option.label}
                    name={String(poll?.id)}
                    onChange={handleAnswer}
                  />
                  <label htmlFor={String(option.id)}>{option.label}</label>
                </span>
              ))}
            </div>
            <p>total number of votes recorded: {total}</p>
          </div>
          <div>
            <Doughnut data={data} />
          </div>
        </DetailContainer>
      </>
    )
  },
)

export default Detail

const DetailContainer = styled(Container)`
  display: flex;
  flex-wrap: wrap;
  background-color: ${props => props.theme.colors.blue};

  @media (min-width: 1140px) {
    div {
      width: 50%;
    }
  }
`

const StyledPublishedDate = styled.div`
  display: flex;
  justify-content: flex-end;
  border-top: 1px solid ${props => props.theme.colors.gray};
  padding-top: 10px;
`
