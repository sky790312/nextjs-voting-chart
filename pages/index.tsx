import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import Head from 'next/head'
import { Container } from '@/GlobalStyles'
import { fetchPolls } from '@/api'
import { PoolSchema } from '@/api/schema'
import { Item } from '@/components/home/Item'
import { TopicSection } from '@/components/home/TopicSection'

export const Home = React.memo(
  (): JSX.Element => {
    const [polls, setPolls] = useState<PoolSchema[]>([])
    const [topicPoll, ...restPolls] = polls

    useEffect(() => {
      const fetchData = async () => {
        try {
          const data = await fetchPolls()
          setPolls(data.polls)
        } catch (error) {
          console.log(error)
        }
      }

      fetchData()
    }, [])

    return (
      <>
        <Head>
          <title>Create Next App</title>
          <link rel='icon' href='/favicon.ico' />
        </Head>
        <Container>
          <TopicSection poll={topicPoll} />
          <PollListContainer>
            {restPolls?.map(poll => (
              <Item key={poll.id} item={poll} />
            ))}
          </PollListContainer>
        </Container>
      </>
    )
  },
)

export default Home

const PollListContainer = styled.div`
  display: flex;
  width: 100%;
  flex-wrap: wrap;

  @media (min-width: 1140px) {
    div {
      width: 50%;
    }
  }
`
