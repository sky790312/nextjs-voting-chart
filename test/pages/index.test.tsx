import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { Home } from '@/pages/index'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<Home />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
