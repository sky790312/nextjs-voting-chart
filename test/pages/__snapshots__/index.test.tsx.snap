// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="GlobalStyles__Container-sc-1m5drgn-0 frTxMm"
  >
    <div
      class="TopicSection__TopicSectionContainer-m3h1n7-0 dlphWy"
    >
      <div>
        <h3>
          Today's Poll
        </h3>
        <p>
          <span>
            ll
          </span>
        </p>
        <div />
        <p>
          total number of votes recorded: 
        </p>
      </div>
      <div>
        <canvas
          height="150"
          width="300"
        />
      </div>
    </div>
    <div
      class="pages__PollListContainer-sc-1cnstz1-0 hbYZaE"
    />
  </div>
</DocumentFragment>
`;
